packages/cli/src/WorkflowRunner.ts

```diff
@@ -564,25 +576,50 @@ export class WorkflowRunner {
 					reject(error);
 				}
 
+				// only pull execution data from the Db when it is needed
+				const executionHasPostExecutionPromises =
+					this.activeExecutions.getPostExecutePromiseCount(executionId) > 0;
+
+				if (executionHasPostExecutionPromises) {
+					Logger.debug(
+						`Reading execution data for execution ${executionId} from db for PostExecutionPromise.`,
+					);
+				} else {
+					Logger.debug(
+						`Skipping execution data for execution ${executionId} since there are no PostExecutionPromise.`,
+					);
+				}
+
 				const fullExecutionData = await Container.get(ExecutionRepository).findSingleExecution(
 					executionId,
 					{
-						includeData: true,
-						unflattenData: true,
+						includeData: executionHasPostExecutionPromises,
+						unflattenData: executionHasPostExecutionPromises,
 					},
 				);
 				if (!fullExecutionData) {
 					return reject(new Error(`Could not find execution with id "${executionId}"`));
 				}
-				const runData = {
-					data: fullExecutionData.data,
+
+				const runData: IRun = {
+					data: {},
 					finished: fullExecutionData.finished,
 					mode: fullExecutionData.mode,
 					startedAt: fullExecutionData.startedAt,
 					stoppedAt: fullExecutionData.stoppedAt,
 				} as IRun;
 
+				if (executionHasPostExecutionPromises) {
+					runData.data = (fullExecutionData as IExecutionResponse).data;
+				}
+
+				// NOTE: for future use, a manual return value via Redis could be used here
+				// if (racingPromisesResult.success && racingPromisesResult.executionResponse) {
+				// 	runData.data = racingPromisesResult.executionResponse.data;
+				// }
+
 				this.activeExecutions.remove(executionId, runData);
+
 				// Normally also static data should be supplied here but as it only used for sending
 				// data to editor-UI is not needed.
 				await hooks.executeHookFunctions('workflowExecuteAfter', [runData]);
```

Comment:
Lucky for us, in this scenario no events rely on `runData.data` but if there was an attached event, this would break. We need to think of a better approach for preventing reading the data. Works for now, since we want to change this.

Reply:
I've added another comment pointing this out to future developers...